rfc: add offset management proposal (#3856)
infinyon/fluvio: rfc/offset-management.md

## Scope

Define a minimal offset management functionality for integration into fluvio.

Use Case 1:
An simple offset retention kind and time should be settable for a given duration and a maximum number of consumers.

> The offsets are managed manually with client side API calls.

Reviewer: ```suggestion
Offsets should be managed as default.   The API for manual control would be provided.
```

Author: Goog suggestion, updated!

Use Case 2:

> Automatic offset commit for single member consumer groups and a default pattern of

Reviewer: Remove this since this implies consumer group

use fluvio::consumer::OffsetManagement;

async fn offset_example() -> Result<()> {
    let client = fluvio::connect();
    let cfg = ConsumerConfigExtBuilder::new()
        .topic("mytopic")
        .partition(0) // optional, defaults to zero?
        .offset_start(Offset::Beginnning)  // optional, defaults to (whatever was before)
        .offset_consumer("my-consumer")    // required id
        .offset_strategy(OffsetManagement::Manual) // optional
        .offset_flush(3000)  // Duration, optional defaults to ?
        .max_bytes(12300) // optional, here as an example of other settings
        .build()?;
    let stream = client.consumer_with_config(cfg);

    // manual offset management is for clients which may not process
    // one record to one offset commit
    while let Some(Ok(record)) = stream.next().await {
        println!("{}", record.get_value().as_utf8_lossy_string());

        if some_condition {
            break;
        }
        stream.offset_commit();
    }

    // synchronously flush for shutdown (or none if intentionally ending processing)
    stream.offset_flush().await?;

    Ok(())
}
```

### Use Case 2: Automatic Offset API

> 

Reviewer: this should go first since automatic should be default

## Scope

Define a minimal offset management functionality for integration into fluvio.

> The initial implementation eg would default for a store of "memory", with a retention of 24 hours, and a count of 100 consumer groups of single consumers. Later expansion could allow for more durable storage (for example, internal partition), higher consumer group counts, multiple member consumer groups, and longer retention times.

Reviewer 2: @digikata - I suggest we remove the default to `memory`, as the whole point of offset management is to make them durable. Also I don't think that a solution that does not recover from durable storage is viable.

> The initial implementation eg would default for a store of "memory", with a retention of 24 hours, and a count of 100 consumer groups of single consumers. Later expansion could allow for more durable storage (for example, internal partition), higher consumer group counts, multiple member consumer groups, and longer retention times.

Reviewer: ```suggestion
Provide a simple minimum way to save and recover offset for the consumer.   We are not considering consumer group use case as part of this RFC
```

> Use Case 1:

Reviewer 2: #### Use Case 1:

> An simple offset retention kind and time should be settable for a given duration and a maximum number of consumers.

Reviewer: ```suggestion
Offset mgt facility would provide way to set TTL for offset lifetime
```

The offsets are managed manually with client side API calls.

This can be used internally for more automatic offset management, but also where the user may need more complex consumption patterns. e.g. consume multiple records at a time before committing an offset.

> Use Case 2:

Reviewer 2: #### Use Case 2:
